### PR TITLE
Fix Illegal Memory Access when fa3 + spec + topk + page_size > 1

### PR DESCRIPTION
**Pull request by yubofredwang**
## Motivation

See: #14624 

In the previous implementation, we did not correctly handle the page_table and lead to IMA because we did not add the first last_partial_page before decode step pages when calculating the expand_metadata's page table. This PR fixes it.

## Modifications

Why we can't simply do:
```
max_pages_expand = decode_length + 2 * self.page_size - 1
page_base = (cache_loc[:, 0] // self.page_size).to(torch.int32)
page_offsets = torch.arange(
      max_pages_expand, device=page_base.device, dtype=torch.int32
)
```
like in PR: #15107 
Because the pages assigned in out_cache_loc is not necessarily continuous chunk. For example, your last extended page is page 1 and the first topk your assigned page is 4. You need to calculate the actual page based on the allocated location.

We need to `// page_size` for each location, and count the unique ones. We have the following cases to handle:

* last_page_len != 0, duplicate. Say page_size = 4, last_page_len = 3, [0, 1, 2] is duplicate. Decode length = 5, out_cache_loc = [3, 4, 5, 6, 7, 8], expected pages = [0, 1, 2], selected indices: [0, 1, 5].
* last_page_len = 0, no duplicate. Say page_size = 4, decode length = 5, out_cache_loc = [4, 5, 6, 7, 8, 9], expected pages = [1, 2, 3].
* Else, pages are not consecutive. Say page_size = 4, decode length = 3, out_cache_loc = [3, 40, 41, 42], expected pages = [0, 10].
* Decode length < page_size. page_size = 4, decode length = 2, last_page_len = 3, out_cache_loc = [3, 4, 5], expected pages = [0, 1].

The solution we are doing here: allocate zeros tensor same size as out_cache_loc, move the unique pages (after // page_size) to the front. We rely on the cache_seq_lens to limit the accessing of the kv cache in a page.

## Accuracy Tests

Added detailed unit tests.

## Benchmarking and Profiling
```
bench_sglang_eagle.py
#questions: 80, Throughput: 1410.20 token/s, Acceptance length: 2.67
```
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 1]
```diff
@@ -1404,23 +1406,12 @@ def init_cuda_graph_state(self, max_bs: int, max_num_tokens: int):
                 ),
                 "page_table": torch.zeros(
                     max_bs * self.topk,
-                    decode_length,
+                    decode_length + 1,  # Additional page for last partial page
                     dtype=torch.int32,
                     device=self.device,
                 ),
```
[the remark this answers is not in the document]
Comment (yubofredwang) [reply python/sglang/srt/layers/attention/flashattention_backend.py:1414]: "The number of unique pages can be up to self.speculative_num_steps" This assumption is wrong. We are doing page_size > 1. So the number of unique pages can only be less.

File: benchmark/mtbench/bench_sglang_eagle.py [hunk 2]
```diff
@@ -38,7 +38,7 @@ def write_answers(filename, model_id, questions, answers):
                 "model_id": model_id,
                 "choices": {
                     "index": 0,
-                    "prompt": [answers[i][0], answers[i][1]],
+                    "turns": [answers[i][0], answers[i][1]],
```
Comment (yubofredwang) [inline benchmark/mtbench/bench_sglang_eagle.py:41]: Reverting the previous wrong change. https://github.com/lm-sys/FastChat/blob/main/fastchat/llm_judge/data/mt_bench/question.jsonl is the right place to download. Not the HF copy.
